// socket-messenger/lib/util/index.js
'use strict';

var is = require('is-type-of');
var debug = require('debug')('socket-msessenger:util');

var INT_32_LENGTH = 4;

var util = {
  getLength(data) {
    var limit = data.readInt32BE(0) + INT_32_LENGTH;
    return limit;
  },

  getCompleteData(data) {
    var limit = util.getLength(data);

    if (data.length < limit) {
      return {};
    }

    var completeData = Buffer.alloc(limit);
    data.copy(completeData, 0, 0, limit);

    var modLength = data.length - limit;
    if (modLength > 0) {
      var modData = Buffer.alloc(modLength);

      data.copy(modData, 0, limit, data.length);
      return {
        completeData,
        modData
      };
    }

    return { completeData };
  },

  encode(data) {
    if (!is.string(data)) {
      data = JSON.stringify(data);
    }

    var bodyLength = util.getBodyLength(data);
    var resultData = Buffer.alloc(INT_32_LENGTH + bodyLength);
    resultData.writeInt32BE(bodyLength, 0);
    resultData.write(data, INT_32_LENGTH);

    return resultData;
  },

  getBodyLength(str) {
    return new Buffer(str).length;
  },

  decode(data) {
    var bodyLength = data.readInt32BE();
    return JSON.parse(data.slice(INT_32_LENGTH, INT_32_LENGTH + bodyLength).toString());
  },

  generateMailNo(prefix, NO) {
    return `${ prefix }#${ NO }`;
  }
};

module.exports = util;
